Add use_args option for explicit helpers

chpatton013 opened the pull request:
Add a new option `use_args` which enables users to pass arbitrary arguments to helpers when they are explicitly specified via `use` for the package-install and upgrade actions.

Closes #38

kewlfft commented on library/aur.py:
```diff
@@ -35,7 +35,6 @@
         description:
             - Whether or not to upgrade whole system.
         type: bool
-        default: no
```
why are you removing this line? default of upgrade is `false`

chpatton013 replied:
I did this to keep consistency with the parameter spec below. Regardless of the outcome of that thread, I can restore the default in the documentation. Let me know what you want to see here.

kewlfft replied:
I would leave the default in the documentation and the readme

kewlfft commented on library/aur.py:
```diff
@@ -213,12 +213,7 @@ def main():
-                'type': 'bool',
-            },
             'upgrade': {
-                'default': False,
```
why are you removing the default here?

chpatton013 replied:
Ah, so this one was tricky!

If you have a default set for an option, the resulting parameter value for that option is set to whatever the default is, even when the option is omitted in the playbook. This means that the builtin validators which reference this option (like our use of `required_one_of` below) will never cause failures - even when no options are passed at all!

However, if you remove this default then the `upgrade` option is omitted from the parameters, and the validators will correctly flag mis-uses of the API.

kewlfft commented on library/aur.py:
```diff
@@ -229,19 +224,27 @@ def main():
         },
+        mutually_exclusive=[['name', 'upgrade']],
         required_one_of=[['name', 'upgrade']],
+        required_by={'use': ['skip_pgp_check', 'ignore_arch']},
```
what does this line achieve?

chpatton013 replied:
This enforces that the `skip_pgp_check` and `ignore_arch` options only be used when `use` is specified, and below we make an assertion that `makepkg` is being used instead of a helper. I might be misinterpreting the intent, but I don't think you want users to be able to specify these options without explicitly saying they intend to use `makepkg`. I don't have a strong feeling about that position, though.

kewlfft commented on library/aur.py:
```diff
@@ -253,13 +256,15 @@ def main():
-    if params['upgrade'] and (params['name'] or use == 'makepkg'):
-        module.fail_json(msg="Upgrade cannot be used with this option.")
+    if params.get('upgrade', False) and use == 'makepkg':
+        module.fail_json(msg="Upgrade cannot be used with the helper 'makepkg'.")
```
I would not call _makepkg_ a helper, helpers are not supported by Arch while _makepkg_ is supported

chpatton013 replied:
Fair enough! What about changing that text to `tool`?

kewlfft replied:
ok

kewlfft commented on library/aur.py:
```diff
@@ -253,13 +256,15 @@ def main():
+
+    if module.check_mode:
+        check_packages(module, params['name'])
+    elif params.get('upgrade', False):
```
why are you using get for upgrade and not for the other booleans?

chpatton013 replied:
With the change discussed above to remove the default for `upgrade`, this property might not be in the `params` dict. So to make sure we handle that default correctly, we could either insert a `False` into the dict if we see that it's not there, or we could use `dict.get` to resolve the default after a failed lookup.

kewlfft replied:
all right

chpatton013 commented:
> I am not familiar with unit tests, I'll take some time to review it, can it be part of another PR ?

Sure, I don't mind splitting that out. Do you mind if I keep the rest of the related changes in here though? That would make things a little easier.

kewlfft commented on README.md:
```diff
@@ -13,14 +13,15 @@ The following helpers are supported and automatically selected, if present, in t
 |aur_only       |yes, **no**                                            |Limit helper operation to the AUR.|
-|skip_pgp_check |yes, **no**                                            |Only valid with makepkg. Skip PGP signatures verification of source file, useful when installing packages without GnuPG properly configured.|
-|ignore_arch    |yes, **no**                                            |Only valid with makepkg. Ignore a missing or incomplete arch field, useful when the PKGBUILD does not have the arch=('yourarch') field.|
+|skip_pgp_check |yes, **no**                                            |Only valid with makepkg. Skip PGP signatures verification of source file, useful when installing packages without GnuPG properly configured. Cannot be used unless use is set to something other than 'auto'.|
```
it should be used only with use set to makepkg actually

kewlfft commented on README.md:
```diff
@@ -13,14 +13,15 @@ The following helpers are supported and automatically selected, if present, in t
-|skip_pgp_check |yes, **no**                                            |Only valid with makepkg. Skip PGP signatures verification of source file, useful when installing packages without GnuPG properly configured.|
-|ignore_arch    |yes, **no**                                            |Only valid with makepkg. Ignore a missing or incomplete arch field, useful when the PKGBUILD does not have the arch=('yourarch') field.|
+|skip_pgp_check |yes, **no**                                            |Only valid with makepkg. Skip PGP signatures verification of source file, useful when installing packages without GnuPG properly configured. Cannot be used unless use is set to something other than 'auto'.|
+|ignore_arch    |yes, **no**                                            |Only valid with makepkg. Ignore a missing or incomplete arch field, useful when the PKGBUILD does not have the arch=('yourarch') field. Cannot be used unless use is set to something other than 'auto'.|
```
same makepkg should be set to be safe

chpatton013 commented:
> thanks for the work, I don't see that `use_args` is used with makepkg. Shall we pass it to makepkg as well?

You're right! Oversight on my part. I meant to do that, and forgot to circle back when I finished everything else. I've fixed this now and added a unit test in the testing branch to account for it.

chpatton013 commented:
> For simplicity I would have seen the extra args as a str, not a list, but it seems there is no rule, some modules use lists some use str, fine with the list. `extra_args` or `extra_opts` however seems to be a more generic option name for this than `use_args`.

That makes sense to me. My initial reasoning was to indicate that the arguments could only be used in tandem with the `use` option, but I think the documentation and error messages do a fine job of explaining that.

As for string vs list, I just think about things in terms of the underlying exec call that is going to be made by the shell, rather than the arg-string that gets passed to the shell. Programmatically, it's easier to work with a list of arguments since you can avoid worrying about the common pitfalls of shell-escaping. But I agree it's simpler for users to specify a simple list.

kewlfft commented on library/aur.py:
```diff
@@ -161,19 +161,20 @@ def install_with_makepkg(module, package):
-        if module.params['ignore_arch']:
-            use_cmd['makepkg'].append('--ignorearch')
-        rc, out, err = module.run_command(use_cmd['makepkg'], cwd=os.path.join(tmpdir, result['Name']), check_rc=True)
+        command = build_command_prefix('makepkg', [], skip_pgp_check=skip_pgp_check, ignore_arch=ignore_arch)
```
I still don't see that `extra_args` is passed to _makepkg_

chpatton013 commented:
Great, let me know if there's anything else you want to see in here.
